// @nestjs/config/dist/config.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigService = void 0;
const common_1 = require("@nestjs/common");
const lodash_get_1 = __importDefault(require("lodash.get"));
const lodash_has_1 = __importDefault(require("lodash.has"));
const lodash_set_1 = __importDefault(require("lodash.set"));
const util_1 = require("util");
const config_constants_1 = require("./config.constants");
let ConfigService = class ConfigService {
    constructor(internalConfig = {}) {
        this.internalConfig = internalConfig;
        this.cache = {};
        this._isCacheEnabled = false;
    }
    get isCacheEnabled() {
        return this._isCacheEnabled;
    }
    set isCacheEnabled(value) {
        this._isCacheEnabled = value;
    }
    get(propertyPath, defaultValue) {
        const validatedEnvValue = this.getFromValidatedEnv(propertyPath);
        if (!util_1.isUndefined(validatedEnvValue)) {
            return validatedEnvValue;
        }
        const processEnvValue = this.getFromProcessEnv(propertyPath, defaultValue);
        if (!util_1.isUndefined(processEnvValue)) {
            return processEnvValue;
        }
        const internalValue = this.getFromInternalConfig(propertyPath);
        if (!util_1.isUndefined(internalValue)) {
            return internalValue;
        }
        return defaultValue;
    }
    getFromCache(propertyPath, defaultValue) {
        const cachedValue = lodash_get_1.default(this.cache, propertyPath);
        return util_1.isUndefined(cachedValue)
            ? defaultValue
            : cachedValue;
    }
    getFromValidatedEnv(propertyPath) {
        const validatedEnvValue = lodash_get_1.default(this.internalConfig[config_constants_1.VALIDATED_ENV_PROPNAME], propertyPath);
        return validatedEnvValue;
    }
    getFromProcessEnv(propertyPath, defaultValue) {
        if (this.isCacheEnabled &&
            lodash_has_1.default(this.cache, propertyPath)) {
            const cachedValue = this.getFromCache(propertyPath, defaultValue);
            return !util_1.isUndefined(cachedValue) ? cachedValue : defaultValue;
        }
        const processValue = lodash_get_1.default(process.env, propertyPath);
        this.setInCacheIfDefined(propertyPath, processValue);
        return processValue;
    }
    getFromInternalConfig(propertyPath) {
        const internalValue = lodash_get_1.default(this.internalConfig, propertyPath);
        return internalValue;
    }
    setInCacheIfDefined(propertyPath, value) {
        if (typeof value === 'undefined') {
            return;
        }
        lodash_set_1.default(this.cache, propertyPath, value);
    }
};
ConfigService = __decorate([
    common_1.Injectable(),
    __param(0, common_1.Optional()),
    __param(0, common_1.Inject(config_constants_1.CONFIGURATION_TOKEN)),
    __metadata("design:paramtypes", [Object])
], ConfigService);
exports.ConfigService = ConfigService;
